src/3rd_party_config.c

```diff
@@ -0,0 +1,173 @@
+#include "3rd_party_internal.h"
+#include "config.h"
+#include "lib/config_parser.h"
+#include "lib/list.h"
+#include "swupd.h"
+#include <stdlib.h>
+
+struct global {
+	void *value1;
+	void *value2;
+};
+
+struct repo {
+	char *repo_name;
+	char *url;
+};
+
+#define MAX_REPO_ALLOWED_FIELDS 2
+#define MAX_GLOBAL_ALLOWED_FIELDS 2
+
+struct list *repos_kv;
+
+bool parse_key_values(char *section, char *key, char *value, void *data)
+{
+	static int repo_field_count = 0;
+	static int global_field_count = 0;
+
+	if (!strncmp(section, "REPO", strlen(section))) {
+		if (key && value) {
+			static struct repo *repo_var;
+			if (repo_field_count == 0) {
+				repo_var = (struct repo *)malloc(sizeof(struct repo));
+			}
+			if (!strncmp(key, "repo-name", strlen(key))) {
+				repo_var->repo_name = strdup_or_die(value);
+				repo_field_count++;
+			}
+
+			if (!strncmp(key, "repo-url", strlen(key))) {
+				repo_var->url = strdup_or_die(value);
+				repo_field_count++;
+			}
+
+			if (repo_field_count == MAX_REPO_ALLOWED_FIELDS) {
+				repos_kv = list_append_data(repos_kv, repo_var);
+				repo_field_count = 0;
+			}
+		}
+		return true;
+	} else if (!strncmp(section, "GLOBAL", strlen(section))) {
+		if (key && value) {
+			static struct global *global_var;
+			if (global_field_count > 2) {
+				return false;
+			}
+
+			if (!strncmp(key, "value1", strlen(key))) {
+				global_var->value1 = strdup_or_die(value);
+				global_field_count++;
+			}
+
+			if (!strncmp(key, "value2", strlen(key))) {
+				global_var->value2 = strdup_or_die(value);
+				global_field_count++;
+			}
+		}
+		return true;
+	}
+	return false;
+}
+
+static inline char *get_repo_config_file()
+{
+	return sys_path_join(globals.state_dir, "/3rd_party/repo.ini");
+}
+
+static int create_repo_config(char *repo_config_file)
+{
+	int fd = open(repo_config_file, O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
+	free_string(&repo_config_file);
```

Comment:
don't free repo_config_file here, free it in the same function where it was allocated